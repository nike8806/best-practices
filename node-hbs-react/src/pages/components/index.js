export { default as WorkEmail } from './WorkEmail/WorkEmail.js';
