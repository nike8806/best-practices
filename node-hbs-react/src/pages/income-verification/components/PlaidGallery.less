.plaid-gallery__bank-logo:hover,
.plaid-gallery__find-bank:hover,
.plaid-gallery__find-bank a:hover {
  background-color: #f9f9f9;
  color: #707884;
  cursor: pointer;
  text-decoration: none;
}

.plaid-gallery__bank-logo {
  background-color: #ffffff;
  border: 1px solid #e2e2e2;
  border-radius: 5px;
  display: block;
  margin: 2px;
  background-position: center;
  background-repeat: no-repeat;
  background-size: 65%;
  height: 115px;
}

.plaid-gallery__find-bank {
  background-color: #ffffff;
  border: 1px solid #e2e2e2;
  border-radius: 5px;
}

.plaid-gallery__bank-logo--chase {
  background-image: url('../images/chase-logo.svg');
}

.plaid-gallery__bank-logo--wells {
  background-image: url('../images/wells-fargo-logo.svg');
}

.plaid-gallery__bank-logo--bofa {
  background-image: url('../images/bofa-logo.svg');
}

.plaid-gallery__bank-logo--pnc {
  background-image: url('../images/pnc-logo.svg');
}

.plaid-gallery__bank-logo--us {
  background-image: url('../images/us-bank-logo.svg');
}

.plaid-gallery__bank-logo--usaa {
  background-image: url('../images/usaa-logo.svg');
  background-size: 50%;
}

@media (min-width: 376px) and (max-width:767px) {
  .plaid-gallery__bank-logo {
      background-size: 40%;
      height: 120px;
  }

  .plaid-gallery__bank-logo--usaa {
      background-size: 26%;
  }
}

@media (min-width: 768px) {
  .plaid-gallery__bank-logo {
      background-size: 55%;
  }

  .plaid-gallery__bank-logo--usaa {
      background-size: 40%;
  }
}
