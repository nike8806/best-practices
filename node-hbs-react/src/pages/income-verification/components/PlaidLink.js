import React, { Component } from 'react';
import PropTypes from 'prop-types';

class PlaidLink extends Component {
  static defaultProps = {
    institution: null,
    onBeforeOpen: null,
    plaidInstance: null,
    className: null
  };

  static propTypes = {
    institution: PropTypes.string,
    onBeforeOpen: PropTypes.func,
    plaidInstance: PropTypes.shape({
      open: PropTypes.func.isRequired
    }),
    children: PropTypes.node,
    className: PropTypes.string
  }

  handleClick = () => {
    const {
      onBeforeOpen,
      institution,
      plaidInstance
    } = this.props;

    if (onBeforeOpen) {
      onBeforeOpen({institution});
    }

    if (plaidInstance) {
      plaidInstance.open(institution);
    }
  }

  render() {
    const {
      children,
      onBeforeOpen, // eslint-disable-line no-unused-vars
      plaidInstance, // eslint-disable-line no-unused-vars
      institution, // eslint-disable-line no-unused-vars
      className,
      ...linkAttributes
    } = this.props;

    return (
      <a
        {...linkAttributes}
        onClick={ this.handleClick }
        className={ className }
      >
        { children }
      </a>
    );
  }
}

export default PlaidLink;
