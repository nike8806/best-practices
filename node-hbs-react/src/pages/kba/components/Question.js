import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Question extends Component {
  static propTypes = {
    type: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    answers: PropTypes.arrayOf(PropTypes.string).isRequired,
    onAnswered: PropTypes.func,
    disabled: PropTypes.bool
  };

  static defaultProps = {
    disabled: false,
    onAnswered: null
  };

  state = {
    // eslint-disable-next-line react/destructuring-assignment, react/no-unused-state
    type: this.props.type,
    answer: null
  };

  static getDerivedStateFromProps(props, state) {
    if (state.type !== props.type) {
      return {
        type: props.type,
        answer: null
      };
    }

    return null;
  }

  /**
   * Sync state with user's selection
   *
   * @param {SyntheticEvent} event The React `SyntheticEvent`
   */
  handleAnswer = (event) => {
    this.setState({ answer: event.target.value });
  };

  /**
   * Invoque prop callback and clear state when clicking Next
   */
  handleNextClick = () => {
    const { onAnswered, type } = this.props;

    if (onAnswered) {
      const { answer } = this.state;
      onAnswered({
        type,
        answer
      });
    }
  };

  /**
   * Render method
   *
   * @return {ReactElement}
   */
  render() {
    const {
      title, type, answers, disabled
    } = this.props;

    const { answer: currentAnswer } = this.state;
    return (
      <div className="question">
        <h5 className="question__title todo-heading">{title}</h5>

        <ul className="question__answers margin-bottom-32 list-unstyled">
          {answers.map(answer => (
            <li key={`${type}-${answer}`} className="question__answer">
              <label
                className="question__answer-text"
                htmlFor={`${type}-${answer}-id`}
              >
                <input
                  id={`${type}-${answer}-id`}
                  type="radio"
                  name="answer"
                  value={answer}
                  className="question__answer-selection todo-radio"
                  checked={currentAnswer === answer}
                  onChange={this.handleAnswer}
                  disabled={disabled}
                />
                <span className="question__answer-radio todo-radio-selector" />
                {answer}
              </label>
            </li>
          ))}
        </ul>

        <button
          type="button"
          className="btn btn-todo center-block question__next"
          disabled={disabled || !currentAnswer}
          onClick={this.handleNextClick}
        >
          Next
        </button>
      </div>
    );
  }
}

export default Question;
