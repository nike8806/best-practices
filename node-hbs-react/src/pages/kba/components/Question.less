.question__answer {
  list-style: none;
  border: 1px solid #eee;
  margin-bottom: 5px;
  background-color: #f5f5f5;
  padding: 0;

  input[type='radio']&-selection {
    margin-right: 8px;
    margin-top: 0;
    vertical-align: middle;

    width: 2em;
    margin: 0;
    padding: 0;
    font-size: 1em;
    opacity: 0;

    display: inline-block;
    margin-left: -2em;
    line-height: 1.5em;
  }
}

.question__answer-text {
  color: #6f7883;
  display: block;
  padding: 15px 15px 15px 10px;
  cursor: pointer;
}
