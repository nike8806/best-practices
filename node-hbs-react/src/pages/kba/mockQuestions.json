{
  "loanId": 127235022,
  "status": true,
  "idNumber": 1944906967,
  "questions": {
    "question": [
      {
        "type": "current.county.b",
        "prompt": "In which county have you lived?",
        "answer": [
          "FULTON",
          "APPANOOSE",
          "LAFAYETTE",
          "None of the above"
        ]
      },
      {
        "type": "previous.address",
        "prompt": "At which of the following addresses have you lived?",
        "answer": [
          "3 CRESSING CT",
          "16015 RACETRACK RD",
          "510 ADAMS RD",
          "None of the above"
        ]
      },
      {
        "type": "purchased.property.from",
        "prompt": "From whom did you purchase the property at 222333 PEACHTREE PLACE?",
        "answer": [
          "A VIRAY",
          "ERIC WALTORS",
          "JOE ANDERSON",
          "None of the above"
        ]
      }
    ]
  }
}
