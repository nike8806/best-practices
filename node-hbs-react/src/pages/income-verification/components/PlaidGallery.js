import React, { Component } from 'react';
import PropTypes from 'prop-types';
import PlaidLink from './PlaidLink';

class PlaidGallery extends Component {
  static propTypes = {
    onBeforeOpen: PropTypes.func,
    plaidInstance: PropTypes.shape({
      open: PropTypes.func.isRequired
    }).isRequired,
    institutions: PropTypes.arrayOf(PropTypes.shape({
      institution: PropTypes.string.isRequired,
      ariaLabel: PropTypes.string.isRequired
    }))
  }

  static defaultProps = {
    onBeforeOpen: null,
    plaidInstance: null,
    institutions: [
      {
        institution: 'chase',
        ariaLabel: 'chase bank logo'
      }, {
        institution: 'wells',
        ariaLabel: 'wells fargo logo'
      }, {
        institution: 'bofa',
        ariaLabel: 'bank of america logo'
      }, {
        institution: 'pnc',
        ariaLabel: 'pnc bank logo'
      }, {
        institution: 'us',
        ariaLabel: 'us bank logo'
      }, {
        institution: 'usaa',
        ariaLabel: 'usaa logo'
      }
    ]
  }

  render() {
    const {
      onBeforeOpen,
      plaidInstance,
      institutions
    } = this.props;

    return (
      <React.Fragment>
        <div className="row-flush">
          { institutions.map(({institution, ariaLabel}) => (
            <div className="col-xs-6 col-sm-4" key={ institution }>
              <PlaidLink
                institution={ institution }
                className={`plaid-gallery__bank-logo plaid-gallery__bank-logo--${institution}`}
                role="img"
                aria-label={ ariaLabel }
                onBeforeOpen={ onBeforeOpen }
                plaidInstance={ plaidInstance }
              />
            </div>))
          }
        </div>

        <div className="plaid-gallery__find-bank margin-top-5 text-center">
          <PlaidLink
            onBeforeOpen={ onBeforeOpen }
            plaidInstance={ plaidInstance }
          >
            <div className="padding-all-10">
              <div className="h4">Find your bank</div>
              <div className="todo-text padding-bottom-0">
                Search 15,000+ banks.
                {' '}
                <span className="visible-sm-inline">Even local banks in your hometown.</span>
              </div>
            </div>
          </PlaidLink>
        </div>
      </React.Fragment>
    );
  }
}

export default PlaidGallery;
