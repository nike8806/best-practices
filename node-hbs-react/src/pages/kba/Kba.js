import React, { Component } from 'react';
import axios from 'lc-axios';
import {
  Timeout, SuccessMessage, Loader, Modal
} from 'components';
import Question from './components/Question';

const DONE_MESSAGES = {
  DONE: 'Thank you. We have verified your identity',
  DONE_NEED_INFORMATION: 'Thank you. We will be in touch if we need additional information'
};

const OUTAGE_MESSAGES = {
  OUTAGE_ERROR: {
    title: 'Sorry, looks like we can\'t do this online',
    subtitle: 'A member of our team will contact you.'
  },
  OUTAGE_QUESTIONS_REQUESTED: {
    title: 'Sorry, looks like we can\'t do this online'
  }
};

/**
 * Handles all interactions with Knowledge Based Authentication UI.
 */
class Kba extends Component {
  state = {
    answers: [],
    questions: [],
    idNumber: null,
    loanId: null,
    currentQuestionIndex: 0,
    // Options: MAIN, DONE, DONE_NEED_INFORMATION,
    // OUTAGE_ERROR, OUTAGE_QUESTIONS_REQUESTED
    showSection: 'MAIN',
    errorMessage: null,
    showLoader: true
  };

  componentDidMount() {
    axios.get('/todo/kba/questions')
      .then((response) => {
        if (response.data.message === 'REQUESTED') {
          this.setState({
            showLoader: false,
            showSection: 'OUTAGE_QUESTIONS_REQUESTED'
          });
        } else {
          this.setState({
            questions: response.data.questions,
            idNumber: response.data.idNumber,
            loanId: response.data.loanId,
            showLoader: false
          });
        }
      })
      .catch(() => {
        this.setState({
          showSection: 'OUTAGE_ERROR',
          showLoader: false
        });
      });
  }

  handleQuestionAnswered = (selection) => {
    this.setState(
      ({ answers, currentQuestionIndex }) => ({
        answers: [...answers, selection],
        currentQuestionIndex: currentQuestionIndex + 1
      }),
      () => {
        const {
          currentQuestionIndex,
          questions
        } = this.state;
        if (currentQuestionIndex >= questions.length) {
          this.setState({
            showLoader: true,
            errorMessage: null
          });

          this.submitAnswers();
        }
      }
    );
  };

  handleDismissModal = () => {
    this.setState({
      errorMessage: ''
    });
  };

  submitAnswers() {
    const {
      answers, loanId, idNumber, currentQuestionIndex
    } = this.state;

    const payload = {
      loanId,
      idquestionnaireAnswers: {
        idNumber
      }
    };

    answers.forEach(({ answer, type }, index) => {
      payload.idquestionnaireAnswers[`question${index + 1}Type`] = type;
      payload.idquestionnaireAnswers[`question${index + 1}Answer`] = answer;
    });

    axios.post('/todo/kba/answers', payload)
      .then((response) => {
        const { data: responseData } = response || {};
        const {
          state: stateResponse = 'COMPLETED'
        } = responseData || {};

        let newState = {};

        switch (stateResponse) {
          case 'DONE':
            newState = {
              questions: [],
              showSection: 'DONE'
            };
            break;

          case 'PENDING':
            newState = {
              questions: responseData.questions,
              currentQuestionIndex: 0,
              answers: [],
              showSection: 'MAIN'
            };
            break;

          case 'COMPLETED':
            newState = {
              questions: [],
              answers: [],
              showSection: 'DONE_NEED_INFORMATION'
            };

            break;
          default:
            newState = {
              errorMessage: 'Sorry, we are having technical difficulties. Please try again later.',
              answers: answers.slice(0, -1),
              currentQuestionIndex: currentQuestionIndex - 1,
              showLoader: false
            };
        }

        this.setState({
          ...newState,
          showLoader: false
        });
      })
      .catch(() => {
        this.setState({
          errorMessage: 'Sorry, we are having technical difficulties. Please try again later.',
          answers: answers.slice(0, -1),
          currentQuestionIndex: currentQuestionIndex - 1,
          showLoader: false
        });
      });
  }

  render() {
    const {
      questions, currentQuestionIndex, errorMessage, showLoader, showSection
    } = this.state;
    const isDone = questions.length > 0 && currentQuestionIndex >= questions.length;
    const question = isDone ? questions[questions.length - 1] : questions[currentQuestionIndex];
    const doneMessage = DONE_MESSAGES[showSection];
    const {
      title: outageTitle,
      subtitle: outageSubtitle
    } = OUTAGE_MESSAGES[showSection] || {};

    return (
      <div className="kba">
        <div className="todo-body">
          {(showSection === 'MAIN') && (
            <React.Fragment>
              <div className="kba__main-section">
                <h1 className="kba__title todo-heading">
                  Answer a few brief questions to keep your account safe
                </h1>

                {question && (
                  <Question
                    type={question.type}
                    title={question.prompt}
                    answers={question.answer}
                    onAnswered={this.handleQuestionAnswered}
                    disabled={isDone}
                  />
                )}
              </div>
            </React.Fragment>
          )}

          {doneMessage && (
            <SuccessMessage
              className="kba__done-section"
              title={doneMessage}
            >
              <a
                className="btn btn-todo"
                href="/account/myAccount.action"
              >
                Return to task list
              </a>
            </SuccessMessage>
          )}

          {outageTitle && (
            <SuccessMessage
              className="kba__outage-section"
              title={outageTitle}
              subtitle={outageSubtitle}
            >
              <a
                className="btn btn-todo"
                href="/account/myAccount.action"
              >
                Return to task list
              </a>
            </SuccessMessage>
          )}

          {errorMessage && (
            <Timeout
              delay={5000}
              onTimeout={this.handleDismissModal}
            >
              <Modal
                className="todo-modal kba__modal-error"
                onDismiss={this.handleDismissModal}
              >
                <h2 className="todo-heading text-center padding-all-0">
                  {errorMessage}
                </h2>
              </Modal>
            </Timeout>
          )}
        </div>

        {showLoader && (
          <Loader theme="fixed" />
        )}
      </div>
    );
  }
}

export default Kba;
