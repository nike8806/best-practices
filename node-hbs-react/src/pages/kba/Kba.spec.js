import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import axios from 'lc-axios';
import Kba from './Kba';
import Question from './components/Question';
import { SuccessMessage, Loader, Modal } from 'components';
import mockQuestions from './mockQuestions.json';
mockQuestions.questions = (mockQuestions.questions && mockQuestions.questions.question) || [];
// TODO: Put this configuration in just one place
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

const questions = mockQuestions.questions;
const mockSelectedAnswers = [
  { type: 'current.county.b', answer: 'FULTON' },
  { type: 'previous.address', answer: '3 CRESSING CT' },
  { type: 'purchased.property.from', answer: 'A VIRAY' }
];

const kbaTitleSelector = '.kba__title';
const outageSelector = '.kba__outage-section';
const mainSectionSelector = '.kba__main-section';
const doneSectionSelector = '.kba__done-section';


describe.only('KBA React component', () => {
  describe('At mouting', () => {
    let component;

    beforeEach(() => {
      component = shallow(<Kba />);
    });

    it('should render correctly', () => {
      expect(component).toMatchSnapshot();
    });

    it('should display a loader indicator', () => {
      expect(component.find(Loader).length).toBe(1);
    });

    it('should trigger an AJAX call to fetch questions', async() => {
      axios.get = jest.fn().mockResolvedValue({});

      await component.instance().componentDidMount();
      expect(axios.get).toBeCalled();
      expect(axios.get).toBeCalledWith('/todo/kba/questions');
      expect(component.find(mainSectionSelector).length).toBe(1);
    });
  });

  describe('At modal oppened', () => {
    let component;
    beforeEach(() => {
      jest.useFakeTimers();
      component = mount(<Kba />);
    });

    it('The modal should be closed when the user click outside', () => {
      component.setState({ errorMessage: 'ERROR' });
      let modalInstance = component.update().find(Modal);
      expect(modalInstance.length).toBe(1);
      modalInstance.find('.modal-backdrop').first().simulate('click', { preventDefault() { } });
      expect(component.update().find(Modal).length).toBe(0);
    });

    it('An ERROR Modal should be closed after 5 seconds', () => {
      component.setState({ errorMessage: 'Has ERROR' });
      expect(component.update().find(Modal).length).toBe(1);
      jest.runAllTimers();
      expect(component.update().find(Modal).length).toBe(0);
    });
  });

  describe('Fetching questions', () => {
    describe('When response with an error', () => {
      let component;

      beforeEach(async() => {
        component = shallow(<Kba />);
        axios.get = jest.fn().mockRejectedValue({ response: {} });

        await component.instance().componentDidMount();
      });

      it('should render correctly', () => {
        component.update();
        expect(component).toMatchSnapshot();
      });

      it('Should render an outage section', () => {
        component.update();
        expect(component.find(Loader).length).toBe(0);
        expect(component.find(outageSelector).length).toBe(1);
        expect(component.find(mainSectionSelector).length).toBe(0);
      });

      it('Should render the kba title on successfully fetching questions', () => {
        expect(component.find(kbaTitleSelector).length).toBe(1);
      });
    });

    describe('With a success response', () => {
      describe('With "REQUESTED" message', () => {
        let component;
        beforeEach(async() => {
          const requestedMock = { data: { ...mockQuestions, message: 'REQUESTED' } };
          axios.get = jest.fn().mockResolvedValue(requestedMock);
          component = mount(<Kba />);

          await component.instance().componentDidMount();
          component.update();
        });

        it('should render correctly', () => {
          expect(component).toMatchSnapshot();
        });

        it('Should render an outage section', () => {
          expect(component.update().find(mainSectionSelector).length).toBe(0);
          expect(component.find(outageSelector).length).toBe(2);
          expect(component.find(doneSectionSelector).length).toBe(0);
          expect(component.find(Modal).length).toBe(0);
        });
      });
      describe('With not "REQUESTED" message', () => {
        let component;
        beforeEach(async() => {
          axios.get = jest.fn().mockResolvedValue({ data: mockQuestions });
          component = mount(<Kba />);

          await component.instance().componentDidMount();
          component.update();
        });

        it('should render correctly', () => {
          expect(component).toMatchSnapshot();
        });

        it('should save the needed part of the  response in a state', () => {
          expect(component.state().questions).toEqual(questions);
        });

        it('should keep track of current question in the state', () => {
          const randomAnsweredQuestions = Math.floor(Math.random() * questions.length) + 1;

          for (let i = 0; i < randomAnsweredQuestions; i++) {
            component.find('input[type="radio"]').at(0).simulate('change');
            component.find('button').simulate('click');
          }

          expect(component.state().currentQuestionIndex).toEqual(randomAnsweredQuestions);
        });

        it('should show the current question data', () => {
          const randomQuestionIndex = Math.floor(Math.random() * questions.length);
          const questionData = questions[randomQuestionIndex];

          component.setState({ currentQuestionIndex: randomQuestionIndex });

          expect(component.find(Question).props()).toMatchObject({
            type: questionData.type,
            title: questionData.prompt,
            answers: questionData.answer
          });
          expect(component.find(mainSectionSelector).length).toBe(1);
          expect(component.find(outageSelector).length).toBe(0);
          expect(component.find(doneSectionSelector).length).toBe(0);
          expect(component.find(Modal).length).toBe(0);
        });

        it('should change the current state between questions until all questions are answered', () => {
          for (let i = 1; i <= questions.length; i++) {
            component.find('input[type="radio"]').at(0).simulate('change');
            component.find('button').simulate('click');

            const currentState = component.state();

            expect(currentState.answers.length).toBe(i);
            expect(currentState.currentQuestionIndex).toBe(i);
          }

          const finalState = component.state();
          expect(finalState.answers.length).toBe(3);
          expect(finalState.currentQuestionIndex).toBe(3);
        });

        it('Should render the kba title while answering the questions', () => {
          expect(component.find(kbaTitleSelector).length).toBe(1);
        });

        it('should trigger a POST AJAX call when all questions are answered', () => {
          const mockPayload = {
            idquestionnaireAnswers: {
              idNumber: 1944906967,
              question1Type: 'current.county.b',
              question1Answer: 'FULTON',
              question2Type: 'previous.address',
              question2Answer: '3 CRESSING CT',
              question3Type: 'purchased.property.from',
              question3Answer: 'A VIRAY'
            },
            loanId: 127235022
          };

          axios.post = jest.fn().mockResolvedValue({});

          for (let i = 0; i < questions.length; i++) {
            component.find('input[type="radio"]').at(0).simulate('change');
            component.find('button').simulate('click');
          }

          expect(component.find(Loader).length).toBe(1);
          expect(axios.post).toHaveBeenCalled();
          expect(axios.post).toHaveBeenCalledTimes(1);
          expect(axios.post).toHaveBeenCalledWith('/todo/kba/answers', mockPayload);
        });
      });
    });
  });
  describe('At sending answers', () => {
    describe('On error sending answers', () => {
      let component;

      beforeEach(() => {
        axios.get = jest.fn().mockResolvedValue({ data: mockQuestions });
        axios.post = jest.fn().mockRejectedValue({ data: {} });

        component = shallow(<Kba />);

        mockSelectedAnswers.forEach(answer => component.instance().handleQuestionAnswered(answer));
      });

      it('should render an error message', () => {
        component.update();

        expect(component.find(Loader).length).toBe(0);
        expect(component.find(Modal).length).toBe(1);
        expect(component.find('.kba__modal-error').length).toBe(1);
        expect(component.update().find(mainSectionSelector).length).toBe(1);
        expect(component.find(outageSelector).length).toBe(0);
        expect(component.find(doneSectionSelector).length).toBe(0);
      });

      it('without clear question', () => {
        component.update();
        expect(component.find(Question).length).toBe(1);
      });

      it('Should render the kba title when an error is thrown', () => {
        expect(component.find(kbaTitleSelector).length).toBe(1);
      });
    });

    describe('All question correctly answered', () => {
      let component;
      const answersResponse = { state: 'DONE' };
      beforeEach(() => {
        axios.get = jest.fn().mockResolvedValue({ data: mockQuestions });
        axios.post = jest.fn().mockResolvedValue({ data: answersResponse});

        component = shallow(<Kba />);

        mockSelectedAnswers.forEach(answer => component.instance().handleQuestionAnswered(answer));
      });

      it('should render a success message', () => {
        component.update();

        expect(component.find(Loader).length).toBe(0);
        expect(component.find(Modal).length).toBe(0);
        expect(component.find(Question).length).toBe(0);
        expect(component.find(SuccessMessage).length).toBe(1);
        expect(component.find(SuccessMessage).props().title).toBe('Thank you. We have verified your identity');
        expect(component.find(doneSectionSelector).length).toBe(1);
        expect(component.update().find(mainSectionSelector).length).toBe(0);
      });

      it('Should not render the kba title if questions are answered and submmited', () => {
        component.update();
        expect(component.state().doneMessage).not.toBeNull();
        expect(component.find(kbaTitleSelector).length).toBe(0);
      });
    });

    describe('Two of tree questions correctly answered', () => {
      let component;
      const answersResponse = {
        state: 'PENDING',
        questions: [{
          answer: [
            'SEBASTIAN',
            'SCHLEICHER',
            'FULTON',
            'None of the above'
          ],
          type: 'current.county.b',
          prompt: 'In which county have you lived?'
        }]
      };

      beforeEach(() => {
        axios.get = jest.fn().mockResolvedValue({ data: mockQuestions });
        axios.post = jest.fn().mockResolvedValue({ data: answersResponse });

        component = shallow(<Kba />);

        mockSelectedAnswers.forEach(answer => component.instance().handleQuestionAnswered(answer));
      });

      it('should set state with the new question', () => {
        component.update();
        expect(component.state().questions).toBe(answersResponse.questions);
        expect(component.state().currentQuestionIndex).toBe(0);
      });

      it('Should render the main Section while the fourth question is being displayed', () => {
        expect(component.find(kbaTitleSelector).length).toBe(1);
        expect(component.find(mainSectionSelector).length).toBe(1);
        expect(component.find(outageSelector).length).toBe(0);
        expect(component.find(doneSectionSelector).length).toBe(0);
        expect(component.find(Modal).length).toBe(0);
      });
    });

    describe('When the answers were wrong', () => {
      let component;
      beforeEach(() => {
        axios.get = jest.fn().mockResolvedValue({ data: mockQuestions });
        axios.post = jest.fn().mockResolvedValue({ data: { state: 'COMPLETED'} } );

        component = shallow(<Kba />);

        mockSelectedAnswers.forEach(answer => component.instance().handleQuestionAnswered(answer));
      });

      it('should render a finished message', () => {
        component.update();

        expect(component.find(Loader).length).toBe(0);
        expect(component.find(Modal).length).toBe(0);
        expect(component.find(Question).length).toBe(0);
        expect(component.find(SuccessMessage).length).toBe(1);
        expect(component.find(SuccessMessage).props().title).toBe('Thank you. We will be in touch if we need additional information');
        expect(component.find(doneSectionSelector).length).toBe(1);
        expect(component.find(outageSelector).length).toBe(0);
        expect(component.find(mainSectionSelector).length).toBe(0);
      });

      it('Should not render the kba title after the questions are answered and submmited', () => {
        component.update();
        expect(component.state().doneMessage).not.toBeNull();
        expect(component.find(kbaTitleSelector).length).toBe(0);
      });
    });
  });
});
