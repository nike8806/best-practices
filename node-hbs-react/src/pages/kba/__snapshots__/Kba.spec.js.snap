// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KBA React component At mouting should render correctly 1`] = `
<div
  className="kba"
>
  <div
    className="todo-body"
  >
    <React.Fragment>
      <div
        className="kba__main-section"
      >
        <h1
          className="kba__title todo-heading"
        >
          Answer a few brief questions to keep your account safe
        </h1>
      </div>
    </React.Fragment>
  </div>
  <Loader
    theme="fixed"
  />
</div>
`;

exports[`KBA React component Fetching questions When response with an error should render correctly 1`] = `
<div
  className="kba"
>
  <div
    className="todo-body"
  >
    <SuccessMessage
      className="kba__outage-section"
      subtitle="A member of our team will contact you."
      title="Sorry, looks like we can't do this online"
      withIcon={false}
    >
      <a
        className="btn btn-todo"
        href="/account/myAccount.action"
      >
        Return to task list
      </a>
    </SuccessMessage>
  </div>
</div>
`;

exports[`KBA React component Fetching questions With a success response With "REQUESTED" message should render correctly 1`] = `
<Kba>
  <div
    className="kba"
  >
    <div
      className="todo-body"
    >
      <SuccessMessage
        className="kba__outage-section"
        title="Sorry, looks like we can't do this online"
        withIcon={false}
      >
        <div
          className="success-message__container text-center kba__outage-section"
        >
          <h1
            className="success-message__title todo-heading big"
          >
            Sorry, looks like we can't do this online
          </h1>
          <div
            className="success-message__content"
          >
            <a
              className="btn btn-todo"
              href="/account/myAccount.action"
            >
              Return to task list
            </a>
          </div>
        </div>
      </SuccessMessage>
    </div>
  </div>
</Kba>
`;

exports[`KBA React component Fetching questions With a success response With not "REQUESTED" message should render correctly 1`] = `
<Kba>
  <div
    className="kba"
  >
    <div
      className="todo-body"
    >
      <div
        className="kba__main-section"
      >
        <h1
          className="kba__title todo-heading"
        >
          Answer a few brief questions to keep your account safe
        </h1>
        <Question
          answers={
            Array [
              "FULTON",
              "APPANOOSE",
              "LAFAYETTE",
              "None of the above",
            ]
          }
          disabled={false}
          onAnswered={[Function]}
          title="In which county have you lived?"
          type="current.county.b"
        >
          <div
            className="question"
          >
            <h5
              className="question__title todo-heading"
            >
              In which county have you lived?
            </h5>
            <ul
              className="question__answers margin-bottom-32 list-unstyled"
            >
              <li
                className="question__answer"
                key="current.county.b-FULTON"
              >
                <label
                  className="question__answer-text"
                >
                  <input
                    checked={false}
                    className="question__answer-selection todo-radio"
                    disabled={false}
                    name="answer"
                    onChange={[Function]}
                    type="radio"
                    value="FULTON"
                  />
                  <span
                    className="question__answer-radio todo-radio-selector"
                  />
                  FULTON
                </label>
              </li>
              <li
                className="question__answer"
                key="current.county.b-APPANOOSE"
              >
                <label
                  className="question__answer-text"
                >
                  <input
                    checked={false}
                    className="question__answer-selection todo-radio"
                    disabled={false}
                    name="answer"
                    onChange={[Function]}
                    type="radio"
                    value="APPANOOSE"
                  />
                  <span
                    className="question__answer-radio todo-radio-selector"
                  />
                  APPANOOSE
                </label>
              </li>
              <li
                className="question__answer"
                key="current.county.b-LAFAYETTE"
              >
                <label
                  className="question__answer-text"
                >
                  <input
                    checked={false}
                    className="question__answer-selection todo-radio"
                    disabled={false}
                    name="answer"
                    onChange={[Function]}
                    type="radio"
                    value="LAFAYETTE"
                  />
                  <span
                    className="question__answer-radio todo-radio-selector"
                  />
                  LAFAYETTE
                </label>
              </li>
              <li
                className="question__answer"
                key="current.county.b-None of the above"
              >
                <label
                  className="question__answer-text"
                >
                  <input
                    checked={false}
                    className="question__answer-selection todo-radio"
                    disabled={false}
                    name="answer"
                    onChange={[Function]}
                    type="radio"
                    value="None of the above"
                  />
                  <span
                    className="question__answer-radio todo-radio-selector"
                  />
                  None of the above
                </label>
              </li>
            </ul>
            <button
              className="btn btn-todo center-block question__next"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              Next
            </button>
          </div>
        </Question>
      </div>
    </div>
  </div>
</Kba>
`;
